---
output:
  bookdown::pdf_document2:
    latex_engine: xelatex
    toc: no
    number_sections: false
    keep_tex: true
    includes:
      in_header: "preamble.tex"
    citation_package: natbib
bibliography: "references.bib"
biblio-style: "genetics"
title: Can we distinguish modes of selective interactions using linkage disequilibrium?
author:
header-includes:
- |
  ```{=latex}
  \usepackage{authblk}
  \author{Aaron P. Ragsdale \\ aaronpeaceragsdale@gmail.com}
  \affil{Department of Human Genetics, McGill University, Montreal, Canada \\
         National Laboratory of Genomics for Biodiversity, Irapuato, Mexico \\
         }
  ```
date: \today
---

```{r setup, echo=FALSE}
```

\begin{abstract} Selection acting on a mutation interferes and interacts with
evolutionary processes at nearby loci, causing allele frequency and correlation
patterns between pairs of selected mutations to deviate from expected
single-locus dynamics. A number of recent studies have used patterns of linkage
disequilibrium between selected variants to test for selective interference and
epistatic interactions, with some disagreement over interpretation of
observations. Interpreting this data is hindered by the relative lack of
analytic or even numerical expectations for patterns of variation between pairs
of loci under the combined effects of selection, dominance, epistasis, and
demography. Here, I develop a numerical approach to compute the expected
two-locus sampling distribution under diploid selection, with arbitrary
epistasis and dominance, and variable recombination and population sizes. I use
this to explore how different models of epistasis and dominance affect expected
signed LD, including for non-steady-state demography relevant to human
populations. Finally, I use whole-genome sequencing data to asses how well we
can differentiate models of selective interference in practice, and find [[that
while xxx, within classes/domains, we see...]]. \end{abstract}

# Introduction

Most new mutations that affect fitness are deleterious and tend to be
eliminated from a population. The average number of generations that a
deleterious mutation segregates in a population depends on the strength of
selection against genomes that carry it, with very damaging mutations kept at
low frequencies in the population and purged relatively rapidly. But in the
time between mutation and fixation or loss, selected variants, both beneficial
and damaging, can dramatically impact patterns of variation in nearby linked
regions. This distortion away from neutral expectations has been well
documented in practically every taxon and population for which we have genomic
data. However, questions remain about the primary mode of interactions between
multiple linked variants and their joint effects on genome-wide patterns of
diversity.

In their foundational paper, @Hill1966-gv recognized that linked negatively
selected variants reciprocally impede the efficacy of population to remove
those mutations. In general, linked selection reduces the fixation probability
of advantageous mutations and increases that of deleterious mutations, compared
to expectations under single-locus dynamics [@Birky1988-jm]. Allele frequency
dynamics of linked selected variants are also predicted to deviate from
expectations without interference. Under a multiplicative fitness model, where
the fitness reduction of a haplotype carrying multiple deleterious variants is
equal to the product of the fitness reduction of each mutation independently,
we should expect to see those mutation segregate on different haplotypes more
often than together, leading to negative, or repulsion, linkage disequilibrium
(LD), although the extent of LD depends non-trivially on the strength of
selection and probability of recombination separating loci [@Hill1966-gv].

Non-additive effects further complicate our evolutionary models, including
dominance (i.e., interactions *within* loci) and epistasis (interactions
*between* loci), and both are thought to be widespread phenomena. A large
fraction of nonsynonymous coding mutations are thought to be at least partially
recessive [@Huber2018-cp; @Agrawal2011-wb], with average levels of dominance
correlating with strength of selection [@Kacser1981-nc]^[correct citation?],
and dominance plays an important role in shaping expected equilibrium allele
frequencies and mutation load of strongly damaging disease mutations
[@Clark1998-kq].^[maybe also cite @Haldane1930-tt] On the other hand, epistasis
differentially impacts deleterious load for asexually vs sexually reproducing
organisms [@Kimura1966-gc; @Kondrashov1995-va] and has been invoked as an
explanation for the evolutionary advantage of sex [@Kondrashov1982-sf;
@Charlesworth1990-kw; @Barton1998-lu], and epistatic interactions may also
drive incompatibilities that lead to postzygotic isolation during the process
of speciation [@Turelli2000-kz]. Within populations, epistasis is known to
cause signed LD to deviate dramatically from zero [@Kondrashov1995-va;
@Charlesworth1990-kw]. However, despite patterns of dominance being appreciated
as important factors shaping linked variation [@Turelli2000-kz; @Zhao2016-bb],
and the evolutionary importance of epistatic interactions, we currently lack
models for predicting correlation patterns between linked mutations under
general selection models.^[See @McVean2000-ox].

In this paper, I develop a numerical approach to solve for the two-locus
sampling distribution under a general diploid selection model with variable
recombination and single-population size history. I use this model to describe
how epistasis and dominance shape expected patterns of signed LD, under both
steady-state and non-equilibrium demography, that have been used to test for
interference and epistasis in population genomic data. I then turn to human
sequencing data and compare patterns of LD for synonymous, missense, and
loss-of-function mutations within and between protein-coding genes and domains,
and assess how well we can expect to discriminate between modes of selective
interactions using genome-wide patterns of LD.

## Empirical observations

The most direct way to test for interactions between linked selected variants
is through deep mutation scanning experiments, in which many distinct mutations
are induced within a target gene and fitness, or some protein function, is
experimentally measured [@Romero2009-yi; @Steinberg2016-is]. For example, using
the model system of the TEM-1 $\beta$-lactamase gene in *E. coli*,
@Bershtein2006-bk found evidence for synergistic epistasis, where multiple
deleterious mutations had a greater effect on fitness than the multiplication
of the observed effects of the individual mutations. The scale of mutation
scanning experiments continues to improve dramatically, promising greater
resolution of the fitness landscape in such model systems.

In most natural populations, directed mutation studies are not possible, and we
must turn to population genetic approaches to infer selective interactions
between observed mutations. Motivated by theory that linked negatively selected
mutations will display negative LD due to interference [@Hill1966-gv], and that
epistasis will drive expected LD away from zero, a number of recent studies
have used patterns of LD within classes of putatively selected variants to
infer modes of selective interactions. In a notable study from
@Callahan2011-ac, pairs of nonsynonymous mutations were found to cluster more
than expected along lineages in the *Drosophilid* species complex, and that
those clustered mutations tended to preserve the charge of the protein and were
in positive LD compared to pairs including synonymous mutations, suggesting
that compensatory nonsynonymous variants were more generally tolerated.

More recently, @Sohail2017-zq observed negative LD between loss-of-function
variants in protein-coding genes (such as stop gains and losses, frameshifts,
and other nonsense mutations) in both human and fruit fly populations, from
which they proposed that widespread synergistic epistasis between these
mutations. Within the past year or two, both @Sandler2020-vk and @Garcia2020-zn
have reevaluated patterns of LD between coding variants in humans, fruit flies,
and *Capsella grandiflora*, and suggested interference and dominance may
instead be driving patterns of LD [@Garcia2020-zn] or questioned whether LD
between loss-of-function variants is significantly different from zero
[@Sandler2020-vk].

A number of factors impede our interpretation of patterns of signed LD between
coding variants. First, for strongly deleterious or loss-of-function variants,
their rarity and low frequency means that statistical measurement of LD and
other diversity measures are quite noisy. Second, comparisons are based on
theory that are confined to limiting and simplistic assumptions, including
steady-state demography, simple selection and interaction models, or unlinked
loci. To generate predictions under more complex models, we rely on expensive
forward simulations. These can be great for building intuition or testing
inference methods, but do not efficiently provide expectations for quantities
of interest across parameter regimes of interest. Theoretical and numerical
studies of haplotype frequencies and LD under general selective interaction
models would be of great benefit.

## Existing theory and numerical methods

Many well-known properties of two-locus dynamics and equilibrium LD come from
early work that centered on the multi-locus diffusion approximation
[@Kimura1955-qe; @Hill1968-vu; @Ohta1969-ie; @Ohta1971-yd]. This includes the
result that genome-wide averages of signed LD are expected to be zero under
neutrality. Under a two-locus biallelic model, where the left locus allows
alleles $A$ and $a$ and the right locus has alleles $B$ and $b$, the standard
covariance measure of LD is defined as $D = f_{AB} - f_{A}f_{B}$ where $f_{AB}$
is the haplotype frequency of double-derived types carrying both $A$ and $B$,
and $f_{A}$ and $f_{B}$ are the marginal frequencies of the derived alleles at
each locus. This covariance decays due to both drift and recombination at a
rate proportional to the inverse of the effective population size and the
distance separating loci: \[\mathbb{E}[D]_{t+1} = \left(1 - \frac{1}{2N_e(t)} -
r \right)\mathbb{E}[D]_t.\] While $\mathbb{E}[D] = 0$, the variance of $D$ is
non-zero, and @Ohta1971-yd presented their famed result that the variance of
$D$ under neutrality and steady-state demography, normalized by the joint
heterozygosity of the two loci, is \[\sigma_d^2 =
\frac{\mathbb{E}[D^2]}{\mathbb{E}[f_A(1-f_A)f_B(1-f_B)]}\approx\frac{5 + \rho /
2}{11 + 13\rho / 2 + \rho^2 / 2},\] where $\rho = 4N_e r$.

Analytic progress beyond these well-known results has come haltingly. In the
1980s, recursions were developed to compute the two-locus sampling distribution
under neutrality, that is, the probability of observing given counts of
two-locus haplotypes in a sample of size $n$ [@Golding1984-pu]. This approach
would continue to be developed and later form the foundation for the inference
of local recombination rates from population genetic data [@Hudson2001-sg;
@McVean2004-gj].

To include selection, however, there have not been many advances beyond the
Monte Carlo simulation approach taken by @Hill1966-gv, albeit with somewhat
more powerful computational resources and more sophisticated software for
performing flexible forward simulation (e.g., @Thornton2014-pn,
@Haller2019-vm). Analytic results for two-locus distributions under selection
are notoriously difficult, with a few notable flashes of progress.
@McVean2007-xd considered the effect of a recent sweep on patterns of LD for
loci nearby or spanning the locus under selection, and in an impressive recent
paper, @Good2020-wj presented analytic solutions for patterns of LD between
rare mutations under additive selection with epistasis. Nonetheless, such
approaches are confined to steady-state demography and constrained selection
models.

Numerical methods straddle the two approaches of expensive and noisy discrete
simulation and analytic solutions, providing a more efficient and practical
method to compute expectations of two-locus diversity measures under a wider
range of parameters and demographic scenarios.^[cite other numerical
approaches, and need to read: @Hallgrimsdottir2008-sq] @Ragsdale2017-gg solved
the two-locus diffusion equation with additive selection at either locus,
although that paper focused on the applicability of two-locus statistics to
demographic inference. More recently, @Ragsdale2019-nt extended the
@Hill1968-vu system for $\mathbb{E}[D^2]$ to compute arbitrary moments of the
distribution of $D$ for any number of populations connected by migration and
admixture. This paper also showed that such a moments-construction can be used
to solve for the two-locus sampling distribution within a single population,
though requires a moment-closure approximation for nonzero recombination and
selection. In the [Methods], I extend this same approach to model arbitrary
diploid selection, which encompasses dominance, epistasis, and other forms of
selective interactions between two loci.

# Methods

## The two-locus sampling distribution with arbitrary selection

The two-locus sampling distribution is the direct analog to the single-locus
site frequency spectrum (SFS) of a given sample size. Instead of describing the
density or number of mutations with a given allele count out of $n$ samples,
the two-locus distribution $\Psi_n$ stores the density or number of pairs of
loci with observed haplotype counts, so that $\Psi_n(i, j, k)$ is the number
of pairs where we see $i$ $AB$ haplotypes, $j$ $Ab$ haplotypes, and $k$ $aB$
haplotypes. Here, we use upper case to denote the derived allele, so the number
of doubly ancestral haplotypes $ab$ is $n-i-j-k$.

- Diffusion equation for $\psi(x, y, z)$
- Sampling to get $\Psi_n$
- Point to @Ragsdale2019-nt supplement, with $\dot{\Psi}_n = \mathcal{D}\Psi_n \ldots$
- Note that the selection operator 

### Drift and recombination

### Selection models with epistasis and dominance

### Moment closure and implementation

### Low-order summaries of the sampling distribution

## Expected LD under inferred human demographic history

## Analysis of human genomic data
- Computing LD statistics commonly used for inference about selection models
  (Figure \@ref(fig:fig1)), using @Ragsdale2020-nz
- Data analysis of Thousand Genomes Project data
  @1000_Genomes_Project_Consortium2015-zq
- Domain information from @Stanek2020-pa

**idea**: LD within genes partitioned by whether two mutations fall within
domains or are in different domains (distance as a proxy? or can we get
information about whether they lie in the same gene or not?)
- see @Yeang2007-gj, @Ivankov2014-tn, @Taverner2020-lk

## Data and software availability

- Thousand genomes VCFs and ancestral sequence
  [@1000_Genomes_Project_Consortium2015-zq]
- Protein domain information [@Stanek2020-pa]
- Implementation as \texttt{moments.TwoLocus} at
  \url{https://bitbucket.org/simongravel/moments}
- Scripts to recreate analyses, figures, and compile this manuscript at
  \url{https://github.com/apragsdale/two_locus_selection}

# Results

## Singed LD under steady-state demography

- Expected LD under models of epistasis and dominance
  at equilibrium (Figures \@ref(fig:fig2) and \@ref(fig:fig3))
  - $\sigma_d^1$, $\sigma_d^2$, $r^2$, for all frequencies
  - The same statistics conditioned on sample allele counts
  - Relationship between "Hudson slice" and allele count-conditioned LD

### Additive selection and epistasis

### Simple dominance within a locus

### General selection and gene-based dominance

## The effect of population size changes on signed LD

- Effect of demography on these same statistics

  - Bottleneck with recovery and expansion
  - Human-like size histories (using @Speidel2019-nj)
  - For each (in own column): \@ref(fig:fig4)
    - Top panel: demography
    - Next panel: additive selection and synergistic epistasis
    - Next panel: dominance with h = 0, 0.2
    - Next panel: gene-based dominance and compensatory mutations

## Signed LD for mutations within and between genes

## Large positive signed LD for missense mutations within annotated protein
domains

- Present data from loss-of-function (LOF), synonymous, and nonsynonymous
  variants from Thousand Genomes, both for pairs of loci within genes and for
  pairs of loci between nearby genes \@ref(fig:fig5). Do we see any
  differences, and do the different patterns in African vs non-African
  populations tell us anything about the possible selection model for different
  categories of variants?

Some results:

- Within genes, both synonymous and missense mutations have LD slightly greater
  than zero, and are not significantly different from each other.
- On the other hand, LOF mutations have negative LD, although the small number
  and low frequency of LOF mutations in human populations means that the
  measurement is quite noisy.
- In fact, taking each population independently, all measurements have confidence
  intervals that overlap with zero
- But assuming independent measurements between populations (which is probably not
  correct, as populations have shared history and thus shared diversity patterns),
  can we say that synonymous and missense mutations have values significantly greater
  than zero and LOF mutations have values less than zero?

Within domains \@ref(fig:fig6):

- Missense and synonymous mutations have positive LD when looking at pairs
  within the same annotated domain.
- Between domains, both are reduced to be roughly zero
- Within domains, missense mutations have signed LD much larger than synonymous
  mutations

# Discussion

- What can we say about epistasis and models of dominance using LD statistics?
  Do we have power to distinguish models using low-order statistics of this
  kind?
- The role of numerical solutions to such models for population genetics
  analyses. They bridge the gap between theory limited by analytic
  intractability and expensive forward simulations such as \texttt{SLiM} and
  \texttt{fwdpy11}.
- Limitations and future directions:
  - multiple populations
  - strong selection leads to numerical instability

# Acknowledgements
- Simon Gravel
- Kevin Thornton

\clearpage
\newpage

# Figures:

```{r fig1, echo=F, fig.cap="(ref:fig1cap)", fig.align="center"}
knitr::include_graphics("figures/fig1.pdf")
```

(ref:fig1cap) \textbf{Additive selection and allele count-conditioned LD.} (A
and B) The distribution of $AB$ haplotype counts in a sample size of 30, in
which we observe 10 $A$ alleles at the left locus, and 10 $B$ alleles at the
right locus. (C and D) The decay of normalized squared LD ($\sigma_d^2$) with
scaled recombination distance for pairs of neutral and selected variants,
respectively. The dashed line is the neutral expectation [@Ohta1971-yd]. (E and
F) Similarly, the decay of $\sigma_d^1=\mathbb{E}[D]/ \mathbb{E}[p(1-p)q(1-q)]$
for neutral and selected variants, respectively. Note the difference in scale
between the two panels.

```{r fig2, echo=F, fig.cap="(ref:fig2cap)", fig.align="center"}
knitr::include_graphics("figures/fig2.pdf")
```

(ref:fig2cap) \textbf{Additive selection with epistasis.} (A) $D^2$, and (B) $D$.

```{r fig3, echo=F, fig.cap="(ref:fig3cap)", fig.align="center"}
knitr::include_graphics("figures/fig3.pdf")
```

(ref:fig3cap) **Dominance effects.** (A) $D^2$, and (B) $D$.

```{r fig4, echo=F, fig.cap="(ref:fig4cap)", fig.align="center"}
knitr::include_graphics("figures/fig4.pdf")
```

(ref:fig4cap) **The effects of demography on signed LD.**

```{r fig5, echo=F, fig.cap="(ref:fig5cap)", fig.align="center"}
knitr::include_graphics("figures/fig5.pdf")
```

(ref:fig5cap) **LD within and between protein-coding genes.**

```{r fig6, echo=F, fig.cap="(ref:fig6cap)", fig.align="center"}
knitr::include_graphics("figures/fig6.pdf")
```

(ref:fig6cap) **LD within and between annotated domains in protein-coding
genes.**


\clearpage
\newpage

# References
